### DISCUSSION TITLE
Alternative way to initialize keycloak

### DISCUSSION THREAD

**Discussion by istvandesign** [dasniko/keycloak-reactjs-demo#30] [Q&A]
Hello and thanks for your tutorials and project example,
I saw this and I am not convinced that this is a good way to initialize authentication in React.
https://github.com/dasniko/keycloak-reactjs-demo/blob/da4fe42a2eb2b2e007cd63d4d0a9dca625fb9a8e/src/index.js#L8

The way I expected this to work is to have a provider HOC that exposes the state in context, not to callback the entire render function. 

Maybe we can come up with a UserServiceProvider so we can use the Provider/Consumer pattern instead of the current example.

I would make the higher order component pass down initialized and user, when the init callback resolves the initialized should be true and user should have the service methods/objects.

**Comment by dasniko** [comment]
You are free to use another approach and to provide a PR. I'll then investigate your suggestion.
But do not introduce another layer of state and/or user object. The Keycloak instance already has all we need!

This way of doing it is for me a working, performant and reliable way and it's proven in various projects in the last (5-6?) years.

**Reply by istvandesign** [reply]
I did not say it's not working, it's just a node-like way of implementation, which looks out of context with the current ways of working with hooks. However this is a much simpler implementation than the AuthContext in react-keycloak.  The one advantage I can clearly think of is that the UserService could be used with any other JS library.

**Reply by dasniko** [reply]
Yes, it's by design and intended. Using React ways of doing something is not always the best thing to do if you can achieve the same easier. As the Keycloak library already does everything we need here, I see no need to use hooks/hocs, as this is just YAGNI.
This would be different when using something else than the Keycloak lib, when you need to implement everything on your own, an AuthContext with all the methods in it would be a possible way.
